### PR TITLE
[New arch] File list isolation

### PR DESCRIPTION
**Pull request by abelgardep**
## Related Issues
App: https://github.com/owncloud/android/issues/2869
_____

### CODE REVIEW

**Review by abelgardep — COMMENTED** [review]
Some questions and suggestions here @fesave @jabarros

File: owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        if (filesToAdd.isNotEmpty()) {
+            files.add(OCFooterFile(manageListOfFilesAndGenerateText(filesToAdd)))
+        }
+
+        diffResult.dispatchUpdatesTo(this)
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
+        return when (viewType) {
+            ViewType.LIST_ITEM.ordinal -> {
```
[the remark this answers is not in the document]
Comment (jabarros) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt:81]: 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/files/filelist/MainFileListFragment.kt [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+
+        fileListOption =
+            if (requireArguments().getParcelable<Parcelable?>(ARG_LIST_FILE_OPTION) != null)
+                requireArguments().getParcelable(ARG_LIST_FILE_OPTION)
+            else
+                FileListOption.ALL_FILES
+
+        if (fileListOption == null) {
+            fileListOption = FileListOption.ALL_FILES
+        }
```
[the remark this answers is not in the document]
Comment (jabarros) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/ui/files/filelist/MainFileListFragment.kt:119]: Much better

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/files/filelist/MainFileListFragment.kt [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+    fun onBrowseUp(): Int {
+        var parentDir: OCFile?
+        var moveCount = 0
+        if (file != null && containerActivity != null) {
+            val storageManager: FileDataStorageManager = containerActivity!!.storageManager
+            var parentPath: String? = null
+            if (file?.parentId != FileDataStorageManager.ROOT_PARENT_ID.toLong()) {
+                parentPath = File(file!!.remotePath).parent
+                parentPath = if (parentPath.endsWith(File.separator)) parentPath else parentPath + File.separator
+                parentDir = storageManager.getFileByPath(parentPath!!)
```
[the remark this answers is not in the document]
Comment (jabarros) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/ui/files/filelist/MainFileListFragment.kt:461]: 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/files/filelist/MainFileListFragment.kt [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+                updateFileListData(filesList = data ?: emptyList())
+            }
+        })
+
+        // Observe the action of retrieving the list of available offline files from DB.
+        mainFileListViewModel.getFilesAvailableOfflineData.observe(viewLifecycleOwner, Event.EventObserver {
+            it.onSuccess { data ->
+                updateFileListData(filesList = data ?: emptyList())
+            }
+        })
```
[the remark this answers is not in the document]
Comment (jabarros) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/ui/files/filelist/MainFileListFragment.kt:214]: You are right, we can observe that one and filter taking into account fileListOption.

File: owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+                val binding = ItemFileListBinding.inflate(LayoutInflater.from(parent.context), parent, false)
+                binding.root.apply {
+                    tag = ViewType.LIST_ITEM
+                    filterTouchesWhenObscured = PreferenceUtils.shouldDisallowTouchesWithOtherVisibleWindows(context)
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt:85]: Done!

File: owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+            }
+        }
+    }
+
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+
+        val viewType = getItemViewType(position)
+
+
+        if (!isFooter(position)) { // Is Item
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt:137]: You're right!  Changed

File: owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+                filterTouchesWhenObscured = PreferenceUtils.shouldDisallowTouchesWithOtherVisibleWindows(context)
+            }
+
+            when (viewType) {
+                ViewType.LIST_ITEM.ordinal -> {
+                    val view = holder as ListViewHolder
+                    view.binding.let {
+                        it.fileListConstraintLayout.apply {
+                            filterTouchesWhenObscured = PreferenceUtils.shouldDisallowTouchesWithOtherVisibleWindows(context)
+                        }
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt:158]: Yes, removed

File: owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+
+                        it.Filename.text = file.fileName
+                        it.fileListSize.text = DisplayUtils.bytesToHumanReadable(file.length, context)
+                        it.fileListLastMod.text = DisplayUtils.getRelativeTimestamp(context, file.modificationTimestamp)
+
+                    }
+                }
+                ViewType.GRID_ITEM.ordinal -> {
+                    //Filename
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt:167]: Understood! I will try to apply it from now on!

File: owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+                }
+
+                ViewType.GRID_IMAGE.ordinal -> {
+                    //SharedIcon
+                    val view = holder as GridImageViewHolder
+                    view.binding.let {
+                        if (file.sharedByLink) {
+                            it.sharedIcon.apply {
+                                setImageResource(R.drawable.ic_shared_by_link)
+                                visibility = View.VISIBLE
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt:179]: Okaaaaaaaay 😆 , it works for me. Changed

File: owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+
+            }
+
+            //TODO Delete it when manage state sync.
+            holder.itemView.findViewById<ImageView>(R.id.localFileIndicator).apply {
+                visibility = View.GONE
+            }
+            holder.itemView.findViewById<ImageView>(R.id.sharedIcon).apply {
+                visibility = View.GONE
+            }
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt:202]: You're right! changed

File: owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+                    }
+
+                    if (file.mimeType == "image/png") {
+                        fileIcon.setBackgroundColor(ContextCompat.getColor(context, R.color.background_color))
+                    }
+                }
+            }
+
+        } else { //Is Footer
+            if (viewType == ViewType.FOOTER.ordinal && !isShowingJustFolders) {
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt:249]: 👍🏻

File: owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+        }
+    }
+
+    private fun manageListOfFilesAndGenerateText(list: List<OCFile>): String {
+        var filesCount = 0
+        var foldersCount = 0
+        val count: Int = list.size
+        var file: OCFile
+        for (i in 0 until count) {
+            file = list[i]
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt:266]: Excelent

File: owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+                }
+            }
+        }
+
+        return generateFooterText(filesCount, foldersCount)
+    }
+
+    private fun generateFooterText(filesCount: Int, foldersCount: Int): String {
+        when {
+            filesCount <= 0 -> {
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/adapters/filelist/FileListAdapter.kt:281]: 👍🏻

File: owncloudApp/src/main/java/com/owncloud/android/presentation/observers/EmptyDataObserver.kt [hunk 14]
```diff
@@ -0,0 +1,36 @@
+package com.owncloud.android.presentation.observers
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/observers/EmptyDataObserver.kt:1]: The missing author 😶‍🌫️

File: owncloudApp/src/main/java/com/owncloud/android/presentation/observers/EmptyDataObserver.kt [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+class EmptyDataObserver constructor(recyclerView: RecyclerView?, emptyView: View?): RecyclerView.AdapterDataObserver() {
+
+    private var emptyView: View? = null
+    private var recyclerView: RecyclerView? = null
+
+    init {
+        this.recyclerView = recyclerView
+        this.emptyView = emptyView
+        checkIfEmpty()
+    }
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/observers/EmptyDataObserver.kt:16]: Changed

File: owncloudApp/src/main/java/com/owncloud/android/presentation/observers/EmptyDataObserver.kt [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+    override fun onChanged() {
+        super.onChanged()
+        checkIfEmpty()
+    }
+
+    override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
+        super.onItemRangeChanged(positionStart, itemCount)
+    }
```
[the remark this answers is not in the document]
Comment (fesave) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/observers/EmptyDataObserver.kt:34]: Deleted

**Pull request merged**